src/flint/types/fmpz_mod_poly.pyx

```diff
@@ -0,0 +1,320 @@
+                )
+            elif typecheck(val[i], fmpz):
+                fmpz_mod_poly_set_coeff_fmpz(
+                    poly, i, (<fmpz>(val[i])).val, self.mod.val
+                )
+            else:
+                val_fmpz = any_as_fmpz(val[i])
+                if val_fmpz is NotImplemented:
+                    fmpz_clear(x)
+                    raise TypeError("unsupported coefficient in list")
+                fmpz_mod_poly_set_coeff_fmpz(
+                    poly, i, (<fmpz>(val_fmpz)).val, self.mod.val
+                )
+        fmpz_clear(x)
+        return 0
+
+    cdef set_any_as_fmpz_mod_poly(self, fmpz_mod_poly_t poly, obj):
+        if typecheck(obj, list):
+            return self.set_list_as_fmpz_mod_poly(poly, obj)
+
+        # Convert fmpz_mod to constant poly
+        if typecheck(obj, fmpz_mod):
+            if self.mod != (<fmpz_mod>obj).ctx:
+                raise ValueError("moduli must match")
+            fmpz_mod_poly_set_fmpz(
+                poly, (<fmpz_mod>obj).val, self.mod.val
+            )
+            return 0
+
+        # Reduced fmpz_poly modulo mod
+        if typecheck(obj, fmpz_poly):
+            fmpz_mod_poly_set_fmpz_poly(
+                poly, (<fmpz_poly>obj).val, self.mod.val
+            )
+            return 0
+
+
+        obj = any_as_fmpz(obj)
+        if obj is NotImplemented:
+            return NotImplemented
+        fmpz_mod_poly_set_fmpz(
+            poly, (<fmpz>obj).val, self.mod.val
+        )
+        return 0
+
+    cdef any_as_fmpz_mod_poly(self, obj):
+        # Convert fmpz_mod_poly
+        if typecheck(obj, fmpz_mod_poly):
+            if self.mod != (<fmpz_mod_poly>obj).ctx:
+                raise ValueError("moduli must match")
+            return obj
+        
+        cdef fmpz_mod_poly res
+        res = fmpz_mod_poly.__new__(fmpz_mod_poly)
+        check = self.set_any_as_fmpz_mod_poly(res.val, obj)
+        if check is NotImplemented:
+            return NotImplemented
+        
+        res.ctx = self
+        return res
+
+    def __eq__(self, other):
+        # Most often, we expect both `fmpz_mod_poly` to be pointing 
+        # to the same ctx, so this seems the fastest way to check
+        if self is other:
+            return True
+        
+        # If they're not the same object in memory, they may have the
+        # same modulus, which is good enough
+        if typecheck(other, fmpz_mod_poly_ctx):
+            return self.mod == (<fmpz_mod_poly_ctx>other).mod
+        return False
+    
+    def __hash__(self):
+        return hash(repr(self))
+
+    def __str__(self):
+        return f"Context for fmpz_mod_poly with modulus: {self.modulus()}"
+
+    def __repr__(self):
+        return f"fmpz_mod_poly_ctx({self.modulus()})"
+
+    def __call__(self, val):
+        return fmpz_mod_poly(val, self)
+
+
+cdef class fmpz_mod_poly(flint_poly):
+    """
+    """
+    def __cinit__(self):
+        fmpz_mod_poly_init(self.val, self.ctx.mod.val)
```

Comment:
Is it safe to use `self.ctx` here?

Reply:
Hmm good point. I'm not sure what the best thing to do here is.

Reply:
I think here `self.ctx` will be None (Cython does that for attributes that are Python objects).

Can there be a private global context for use here?

We either need to do like `fmpz_mod_ctx` or perhaps there could be a global `fmpz_mod_ctx_t` to use in these situations.

Reply:
We could have a global `fmpz_mod_ctx_t` which is initalised and set to have modulus `1`? I still need to address this in the PR

Reply:
For `fmpz_mod_ctx` we don't need to worry, as the init only sets the value, and not the context (the context here is a python type, not a c-type)

I have included the above in the latest commit